x/examples/soax-relay/main.go

```diff
@@ -0,0 +1,260 @@
+}
+
+// Valid checks if the provided prefix and password are valid
+func (s *StaticCredentialStore) Valid(prefix, password, userAddr string) bool {
+	storedPassword, exists := s.credentials[prefix]
+	return exists && storedPassword == password
+}
+
+func udpAssociateHandler(ctx context.Context, writer io.Writer, request *socks5.Request) error {
+	// Extract the suffix from the auth context
+	suffix, ok := request.AuthContext.Payload["suffix"]
+	if !ok {
+		return errors.New("no suffix found in auth context")
+	}
+
+	soaxPackage := viper.GetString("soax.package")
+	soaxPassword := viper.GetString("soax.password")
+	soaxAddress := viper.GetString("soax.address")
+
+	// Construct the upstream soax username by concatenating
+	// the soax package name with the suffix
+	soaxUsername := soaxPackage + suffix
+	log.Printf("Sending associate command with username: %s\n", soaxUsername)
+
+	streamEndpoint := transport.StreamDialerEndpoint{Dialer: &transport.TCPDialer{}, Address: soaxAddress}
+	client, err := csocks5.NewClient(&streamEndpoint)
+	if err != nil {
+		return err
+	}
+
+	err = client.SetCredentials([]byte(soaxUsername), []byte(soaxPassword))
+	if err != nil {
+		return err
+	}
+
+	client.EnablePacket(&transport.UDPDialer{})
+
+	conn, bindAddr, err := client.ConnectAndRequest(ctx, csocks5.CmdUDPAssociate, "0.0.0.0:0")
+	if err != nil {
+		return err
+	}
+
+	// Start a goroutine to close the connection after a timeout
+	go func() {
+		timeout := viper.GetDuration("udp_timeout")
+		<-time.After(timeout)
+		log.Println("Closing UDP associate connection after timeout")
+		conn.Close()
+	}()
+
+	if err = socks5.SendReply(writer, statute.RepSuccess, convertToNetAddr("udp", bindAddr.IP, bindAddr.Port)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func convertToNetAddr(network string, ip netip.Addr, port uint16) net.Addr {
+	// Convert netip.Addr to net.IP
+	netIP := ip.AsSlice()
+	switch network {
+	case "tcp", "tcp4", "tcp6":
+		// Create a net.TCPAddr
+		tcpAddr := &net.TCPAddr{
+			IP:   netIP,
+			Port: int(port),
+		}
+		return tcpAddr
+	case "udp", "udp4", "udp6":
+		// Create a net.UDPAddr
+		udpAddr := &net.UDPAddr{
+			IP:   netIP,
+			Port: int(port),
+		}
+		return udpAddr
+	}
+	return nil
+}
+
+func loadConfig() error {
+	viper.SetConfigName("config")
+	viper.SetConfigType("yaml")
+	viper.AddConfigPath(".")
+
+	if err := viper.ReadInConfig(); err != nil {
+		return fmt.Errorf("failed to read config file: %w", err)
+	}
+	return nil
+}
+
+func main() {
+
+	if err := loadConfig(); err != nil {
+		log.Fatalf("Failed to load config: %v", err)
+	}
+
+	creds := &StaticCredentialStore{
+		credentials: viper.GetStringMapString("credentials"),
+	}
+
+	// Create the custom authenticator
+	customAuth := CustomAuthenticator{
+		PrefixCredentials: creds,
+	}
+
+	server := socks5.NewServer(
+		socks5.WithAuthMethods([]socks5.Authenticator{customAuth}),
+		socks5.WithLogger(socks5.NewLogger(log.New(os.Stdout, "socks5: ", log.LstdFlags))),
```

Comment:
Not required, but you could use slog with https://pkg.go.dev/log/slog#NewLogLogger